Edit: repeat code

sazk07 opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

<!-- ✍️ Summarize your changes in one or two sentences -->
This commit will extract out repeat code into a new function and call that function from the event handlers. This will improve code readability.

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

sazk07 committed:
Edit: repeat code

This commit will extract out repeat code into a new function and call that function from the event handlers. This will improve code readability.

chrisdavidmills committed:
Merge branch 'main' into patch-1

sazk07 committed:
Edit: variable scope

This commit will tightly scope variable because it is not referenced anywhere else outside the function.

chrisdavidmills reviewed (commented):
Hi there @sazk07, and thank you for your work on this.

I think what you are trying to do here is good, however, on trying the example out, I was concerned that the validity checking doesn't really work properly (not your fault, you are just using what is already there).

I think the main issue is that the `isValid` check is broken. It definitely shouldn't be returning `true` if the email input has an empty value. 

Other issues:

- The `error` `className` setting code is weird. That span's job is to report an error, and a bunch of the styling relies on the class, so to set and unset it is strange. An ID would be much better.
- Really, the reporting of the email as valid or invalid should be a single function, and the showing or hiding or the error message on submit should be a different function. They are different concerns.
- Removing the outline on focus is problematic, accessibility-wise.

I played around with the code a bit, and ended up with this for the HTML (I don't think this is changed much except for the `id` on the second span):

```html
<form>
  <p>
    <label for="mail">
      <span>Please enter an email address:</span>
      <input type="text" id="mail" name="mail" />
      <span id="error" aria-live="polite"></span>
    </label>
  </p>
  <button>Submit</button>
</form>  
```

This for the CSS (updated `class` to `id` for the error span, added a style change for the input on focus, simplified the selectors to get rid of specificity issues):

```css
body {
  font: 1em sans-serif;
  width: 200px;
  padding: 0;
  margin: 0 auto;
}

form {
  max-width: 200px;
}

p * {
  display: block;
}

input {
  appearance: none;
  width: 100%;
  border: 1px solid #333;
  margin: 0;

  font-family: inherit;
  font-size: 90%;

  box-sizing: border-box;
}

/* invalid fields */
input.invalid {
  border: 2px solid #900;
  background-color: #fdd;
}

input:focus.invalid {
  outline: none;
  /* make sure keyboard-only users see a change when focusing */
  border-style: dashed;
}

/* error messages */
#error {
  width: 100%;
  font-size: 80%;
  color: white;
  background-color: #900;
  border-radius: 0 0 5px 5px;
  box-sizing: border-box;
}

.active {
  padding: 0.3rem;
}
```

And this for the JavaScript (I feel like this solves the problems outlined above, and is a bit easier to understand):

```js
const form = document.querySelector("form");
const email = document.getElementById("mail");
const error = document.getElementById("error");

// As per the HTML Specification
const emailRegExp =
  /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

// Create a function that checks if the email is valid, returning a boolean
function isValidEmail() {
  let isValid;
  if(email.value.length === 0) {
    isValid = false;
  } else if(!emailRegExp.test(email.value)) {
    isValid = false;
  } else {
    isValid = true;
  }
  
  // Set class name to indicate validity
  email.className = isValid ? "valid" : "invalid";

  return isValid;
};

// Show or hide the error message based on the validity state
function updateError(isValid) {
  if(!isValid) {
    error.textContent = "I expect an email, darling!";
    error.className = "active";
  } else {
    error.textContent = "";
    error.className = "";
  }
};

// Now we can rebuild our validation constraint
// Because we do not rely on CSS pseudo-class, we have to
// explicitly set the valid/invalid class on our email field

window.addEventListener("load", () => {
  // Here, we test if the field is empty (remember, the field is not required)
  // If it is not, we check if its content is a well-formed email address.
  isValidEmail();
});

// This updates the validity styling when the user types in the field
email.addEventListener("input", () => {
  isValid = isValidEmail();

  // Check if the error is being shown, and hide it as the input value is updated
  // We don't really want it being shown except on submit
  if(error.className === "active") {
    error.textContent = "";
    error.className = "";
  }
});

// This updates the validity styling when the user tries to submit the data,
// And shows or hides the error message as required
form.addEventListener("submit", (event) => {
  event.preventDefault();

  const isValid = isValidEmail();
  updateError(isValid);
});
```

Sorry if this review is so long! Let me know what you think of my code suggestion, and feel free to use as much of it as you want. I think you should at least get the validation working properly before we merge this.

Thanks again.

sazk07 committed:
Refactor: functions for readability and separation of concerns

This commit will refactor functions with the goal of being readable for viewers.
This commit will also separate functions that check if input is of valid
email type and functions that will set the email class name based on that validation.
This commit will also set and/or remove error class name based on the value of email validation function.

sazk07 committed:
Refactor: functions for readability and separation of concerns

This commit will refactor functions with the goal of being readable for viewers.
This commit will also fix validation result when email value length is 0.
This commit will also separate functions that check if input is of valid
email type and functions that will set the email class name based on
that validation.
This commit will also set and/or remove error class name based on the value of email validation function.

sazk07 commented:
> Hi there @sazk07, and thank you for your work on this.
> 
> I think what you are trying to do here is good, however, on trying the example out, I was concerned that the validity checking doesn't really work properly (not your fault, you are just using what is already there).
> 
> I think the main issue is that the `isValid` check is broken. It definitely shouldn't be returning `true` if the email input has an empty value.
> 
> Other issues:
> 
>     * The `error` `className` setting code is weird. That span's job is to report an error, and a bunch of the styling relies on the class, so to set and unset it is strange. An ID would be much better.
> 
>     * Really, the reporting of the email as valid or invalid should be a single function, and the showing or hiding or the error message on submit should be a different function. They are different concerns.
> 
>     * Removing the outline on focus is problematic, accessibility-wise.
> 
> 
> I played around with the code a bit, and ended up with this for the HTML (I don't think this is changed much except for the `id` on the second span):
> 
> ```
> <form>
>   <p>
>     <label for="mail">
>       <span>Please enter an email address:</span>
>       <input type="text" id="mail" name="mail" />
>       <span id="error" aria-live="polite"></span>
>     </label>
>   </p>
>   <button>Submit</button>
> </form>  
> ```
> 
> This for the CSS (updated `class` to `id` for the error span, added a style change for the input on focus, simplified the selectors to get rid of specificity issues):
> 
> ```css
> body {
>   font: 1em sans-serif;
>   width: 200px;
>   padding: 0;
>   margin: 0 auto;
> }
> 
> form {
>   max-width: 200px;
> }
> 
> p * {
>   display: block;
> }
> 
> input {
>   appearance: none;
>   width: 100%;
>   border: 1px solid #333;
>   margin: 0;
> 
>   font-family: inherit;
>   font-size: 90%;
> 
>   box-sizing: border-box;
> }
> 
> /* invalid fields */
> input.invalid {
>   border: 2px solid #900;
>   background-color: #fdd;
> }
> 
> input:focus.invalid {
>   outline: none;
>   /* make sure keyboard-only users see a change when focusing */
>   border-style: dashed;
> }
> 
> /* error messages */
> #error {
>   width: 100%;
>   font-size: 80%;
>   color: white;
>   background-color: #900;
>   border-radius: 0 0 5px 5px;
>   box-sizing: border-box;
> }
> 
> .active {
>   padding: 0.3rem;
> }
> ```
> 
> And this for the JavaScript (I feel like this solves the problems outlined above, and is a bit easier to understand):
> 
> ```js
> const form = document.querySelector("form");
> const email = document.getElementById("mail");
> const error = document.getElementById("error");
> 
> // As per the HTML Specification
> const emailRegExp =
>   /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
> 
> // Create a function that checks if the email is valid, returning a boolean
> function isValidEmail() {
>   let isValid;
>   if(email.value.length === 0) {
>     isValid = false;
>   } else if(!emailRegExp.test(email.value)) {
>     isValid = false;
>   } else {
>     isValid = true;
>   }
>   
>   // Set class name to indicate validity
>   email.className = isValid ? "valid" : "invalid";
> 
>   return isValid;
> };
> 
> // Show or hide the error message based on the validity state
> function updateError(isValid) {
>   if(!isValid) {
>     error.textContent = "I expect an email, darling!";
>     error.className = "active";
>   } else {
>     error.textContent = "";
>     error.className = "";
>   }
> };
> 
> // Now we can rebuild our validation constraint
> // Because we do not rely on CSS pseudo-class, we have to
> // explicitly set the valid/invalid class on our email field
> 
> window.addEventListener("load", () => {
>   // Here, we test if the field is empty (remember, the field is not required)
>   // If it is not, we check if its content is a well-formed email address.
>   isValidEmail();
> });
> 
> // This updates the validity styling when the user types in the field
> email.addEventListener("input", () => {
>   isValid = isValidEmail();
> 
>   // Check if the error is being shown, and hide it as the input value is updated
>   // We don't really want it being shown except on submit
>   if(error.className === "active") {
>     error.textContent = "";
>     error.className = "";
>   }
> });
> 
> // This updates the validity styling when the user tries to submit the data,
> // And shows or hides the error message as required
> form.addEventListener("submit", (event) => {
>   event.preventDefault();
> 
>   const isValid = isValidEmail();
>   updateError(isValid);
> });
> ```
> 
> Sorry if this review is so long! Let me know what you think of my code suggestion, and feel free to use as much of it as you want. I think you should at least get the validation working properly before we merge this.
> 
> Thanks again.

@chrisdavidmills I see what you mean. I have updated the code to check if email value length is 0 to return false. The original author may have written that condition to 'reset' the input element's attributes on window load event.

The new commit will refactor code to separate the email validation function and the function to set the className of the input element. This helps with code readability in the caller where these functions are called. Previously, embedding the email className setter within the email validation checking function was confusing.

The new commit will remove error className instead of setting the error className as an empty string in the updateError() function.

Overall, the new commit will improve code readability for anyone reading the Client-side Form Validation section of Learn Web Development.

chrisdavidmills committed:
Merge branch 'main' into patch-1

chrisdavidmills commented on files/en-us/learn_web_development/extensions/forms/form_validation/index.md:
```diff
@@ -784,49 +784,69 @@ The big changes are in the JavaScript code, which needs to do much more heavy li
-const emailRegExp =
-  /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
+// Check if the email is valid
+const isValidEmail = () => {
```
This should be declared outside the function. It's inefficient to declare it every time the function is run.

chrisdavidmills commented on files/en-us/learn_web_development/extensions/forms/form_validation/index.md:
```diff
@@ -714,9 +714,9 @@ The HTML is almost the same; we just removed the HTML validation features.
-      <span class="error" aria-live="polite"></span>
     </label>
+    <input type="text" id="mail" name="mail" />
+    <span class="error" aria-live="polite"></span>
```
```suggestion
    <span id="error" aria-live="polite"></span>
```

This change is needed, otherwise `error` is null and `updateError()` doesn't work.

chrisdavidmills reviewed (changes requested):
Hello again @sazk07! This is much better, nice work. I just had a couple of small comments for you to look at. I think I've spotted all the errors, but make sure it all works before your next submission.

Cheers!

sazk07 committed:
Fix: email regex and error element attr

This commit will change the attribute of error span element from className to id.
This commit will also declare the email regex pattern outside the check email validation function.

sazk07 committed:
Edit: formatting whitespace

sazk07 committed:
Edit: formatting suggestion by linter/formatter

sazk07 reviewed (commented):
@chrisdavidmills Requested changes made

chrisdavidmills committed:
Merge branch 'main' into patch-1

chrisdavidmills committed:
Update index.md

chrisdavidmills committed:
Update index.md

chrisdavidmills reviewed (approved):
Thanks again @sazk07!

I tried this locally, and it all works fine, however the submit handler doesn't work on the in-page live example. I'm not sure exactly why, but I decided to make a couple of small changes to get this over the finish line so we can merge it:

- I updated the `handleInput()` function so it basically does the same thing as `handleSubmit()`, but without the `preventDefault()` call. The result is that the message now appears and disappears on keystrokes, but at least you get to see what it would look like.
- I increased the height of the live sample `<iframe>` so it is tall enough that you don't get a scrollbar when the message appears.

I hope that's OK. Thanks again for all the great work!
